File: app/src/main/java/com/example/calculatorapp/domains/calculator/internals/adapters/ExpressionEvaluator.java
package com.example.calculatorapp.domains.calculator.internals.adapters;

import static com.example.calculatorapp.domains.calculator.internals.checkers.CalculatorChecker.INSTANCE;

import com.example.calculatorapp.domains.calculator.internals.errors.NotValidExpressionException;

import net.objecthunter.exp4j.ExpressionBuilder;

public class ExpressionEvaluator {
    public static String getEvaluatedCalculationExpression(String calculationExpression) throws NotValidExpressionException {
        try {
            double evaluatedCalculationExpression = new ExpressionBuilder(calculationExpression)
                    .build()
                    .evaluate();

            if (INSTANCE.isCalculationExpressionRationalNumber(evaluatedCalculationExpression)) {
                return Double.toString(evaluatedCalculationExpression);
            } else {
                return Integer.toString((int) evaluatedCalculationExpression);
            }
        } catch (Exception exception) {
            throw new NotValidExpressionException();
        }
    }
}

File: app/src/main/java/com/example/calculatorapp/domains/calculator/infrastructure/anticorruption_layer/ExpressionEvaluator.kt
package com.example.calculatorapp.domains.calculator.infrastructure.anticorruption_layer

import com.example.calculatorapp.constants.domains.CalculatorConstants.Companion.NOT_VALID_EXPRESSION_EXCEPTION_MESSAGE
import com.example.calculatorapp.domains.calculator.infrastructure.specifications.CalculatorSpecifications.isCalculationExpressionRationalNumber
import net.objecthunter.exp4j.ExpressionBuilder

class ExpressionEvaluator {
    companion object {
        fun getEvaluatedCalculationExpression(calculationExpression: String): String {
            try {
                val evaluatedCalculationExpression = ExpressionBuilder(calculationExpression)
                    .build()
                    .evaluate()

                return if (isCalculationExpressionRationalNumber(
                        evaluatedCalculationExpression
                    )
                )
                    evaluatedCalculationExpression.toString() else
                    evaluatedCalculationExpression.toInt()
                        .toString()
            } catch (exception: Exception) {
                return NOT_VALID_EXPRESSION_EXCEPTION_MESSAGE
            }
        }
    }
}